src/core/net/dnssd_server.cpp

```diff
@@ -921,6 +952,154 @@ bool Server::Response::QueryNameMatchesService(const Srp::Server::Service &aServ
 
 #endif // OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
 
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE || OPENTHREAD_CONFIG_DNSSD_DISCOVERY_PROXY_ENABLE
+
+void Server::ConstructSoaServerName(void)
+{
+    if (mSoaServerName[0] == kNullChar)
+    {
+        StringWriter writer(mSoaServerName, sizeof(mSoaServerName));
+
+        writer.Append("otDNS%s", Get<Mac::Mac>().GetExtAddress().ToString().AsCString());
+    }
+}
+
+Error Server::Response::ResolveSoaOrNsQuery(void)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(QueryNameIsForDomain(kDefaultDomainName));
+
+    mSection = kAnswerSection;
+
+    // Handle AAAA (or ANY) query for the server host name used for
+    // SOA/NS answer.
+
+    if (mQuestions.IsFor(kRrTypeAaaa) || mQuestions.IsFor(kRrTypeAny))
+    {
+        Name::Buffer serverFullName;
+
+        ConstructFullName(Get<Server>().mSoaServerName, serverFullName);
+
+        if (QueryNameMatches(serverFullName))
+        {
+            for (Ip6::Netif::UnicastAddress &unicastAddr : Get<ThreadNetif>().GetUnicastAddresses())
+            {
+                if (!unicastAddr.mValid || !unicastAddr.mPreferred || unicastAddr.GetAddress().IsLinkLocalUnicast() ||
+                    Get<Mle::Mle>().IsMeshLocalAddress(unicastAddr.GetAddress()))
+                {
+                    continue;
+                }
+
+                SuccessOrExit(error = AppendAaaaRecord(unicastAddr.GetAddress(), kServerAaaaTtl));
+            }
+
+            if (IsEmpty())
+            {
+                SuccessOrExit(error = AppendAaaaRecord(Get<Mle::Mle>().GetMeshLocalEid(), kServerAaaaTtl));
+            }
+
+            ExitNow();
+        }
+    }
+
+    // Handle SOA or NS (or ANY) queries.
+
+    if (mQuestions.IsFor(kRrTypeSoa) || mQuestions.IsFor(kRrTypeNs) || mQuestions.IsFor(kRrTypeAny))
+    {
+        if (!QueryNameMatches(kDefaultDomainName))
+        {
+            // The SOA and NS query should be for the domain name
+            // (`default.service.arpa`). If it is for a name under the
+            // default domain, e.g. `somehost.default.service.arpa`, we
+            // still provide the info but in Authority section
+            // unless the query is for ANY record type.
+
+            VerifyOrExit(!mQuestions.IsFor(kRrTypeAny));
+            mSection = kAuthoritySection;
+        }
+
+        if (mQuestions.IsFor(kRrTypeSoa) || mQuestions.IsFor(kRrTypeAny))
+        {
+            SuccessOrExit(error = AppendSoaRecord());
+        }
+
+        if (mQuestions.IsFor(kRrTypeNs) || mQuestions.IsFor(kRrTypeAny))
+        {
+            SuccessOrExit(error = AppendNsRecord());
```

Comment:
Looking at the SPEC-1389 'mini-spec' point 6: it was intended to be fully identical to point 3, so to send a SOA record in the Authority section (not NS).  Trying to list all the cases here:
* Query ANY somename.default.service.arpa -> (execution doesn't continue to here, correct)
* Query SOA somename.default.service.arpa -> adds SOA in Authority section, correct
* Query NS   somename.default.service.arpa -> *should add SOA in Authority section*
* Query ANY default.service.arpa -> adds SOA and NS records in Answer section, correct
* Query SOA default.service.arpa -> adds SOA in Answer section, correct
* Query NS   default.service.arpa -> adds NS in Answer section, correct

Reply:
Thanks @EskoDijk.

Changed this in new push.